src/components/engagement/engagement.service.ts

```diff
@@ -15,54 +26,823 @@ import {
+          ...ceremony,
+        },
+        canRead: !!result.canReadCeremony,
+        canEdit: !!result.canEditCeremony,
+      },
+      completeDate: {
+        value: result.completeDate,
+        canRead: !!result.canReadCompleteDate,
+        canEdit: !!result.canEditCompleteDate,
+      },
+      disbursementCompleteDate: {
+        value: result.disbursementCompleteDate,
+        canRead: !!result.CanReadDisbursementCompleteDate,
+        canEdit: !!result.CanEditDisbursementCompleteDate,
+      },
+      communicationsCompleteDate: {
+        value: result.communicationsCompleteDate,
+        canRead: !!result.canReadCommunicationsCompleteDate,
+        canEdit: !!result.canEditCommunicationsCompleteDate,
+      },
+      modifiedAt: result.modifiedAt,
+      startDate: {
+        value: result.startDate,
+        canRead: !!result.canReadStartDate,
+        canEdit: !!result.canEditStartDate,
+      },
+      endDate: {
+        value: result.endDate,
+        canRead: !!result.canReadEndDate,
+        canEdit: !!result.canEditEndDate,
+      },
+      initialEndDate: {
+        value: result.initialEndDate,
+        canRead: !!result.canReadInitialEndDate,
+        canEdit: !!result.canEditInitialEndDate,
+      },
+      lastSuspendedAt: {
+        value: result.lastSuspendedAt,
+        canRead: !!result.canReadLastSuspendedAt,
+        canEdit: !!result.canEditLastSuspendedAt,
+      },
+      lastReactivatedAt: {
+        value: result.lastReactivatedAt,
+        canRead: !!result.canReadLastReactivatedAt,
+        canEdit: !!result.canEditLastReactivatedAt,
+      },
+      statusModifiedAt: {
+        value: result.statusModifiedAt,
+        canRead: !!result.canReadStatusModifiedAt,
+        canEdit: !!result.canEditStatusModifiedAt,
+      },
+    };
   }
 
   async list(
-    _input: EngagementListInput,
-    _session: ISession
+    { page, count, sort, order, filter }: EngagementListInput,
+    session: ISession
   ): Promise<EngagementListOutput> {
-    throw new NotImplementedException();
+    const matchNode =
+      filter.type === 'internship'
+        ? 'internship:InternshipEngagement'
+        : filter.type === 'language'
+        ? 'language:LanguageEngagement'
+        : 'engagement';
+
+    const tmpNode = matchNode.substring(0, matchNode.indexOf(':'));
+    const node = tmpNode ? tmpNode : 'engagement';
+
+    const query = `
+      MATCH (${matchNode} {active: true})<-[:engagement {active: true}]-(project)
+      RETURN ${node}.id as id
+      ORDER BY ${node}.${sort} ${order}
+      SKIP $skip LIMIT $count
+    `;
+    const result = await this.db
+      .query()
+      .raw(query, {
+        skip: (page - 1) * count,
+        count,
+        type: filter.type,
+      })
+      .run();
+
+    const items = await Promise.all(
+      result.map(row => this.readOne(row.id, session))
+    );
+
+    return {
+      items,
+      total: items.length,
+      hasMore: false,
+    };
   }
 
   async listProducts(
```

Comment:
This doesn't filter by engagement yet or check for permissions

Reply:
1. It is filtered by `internship`, `language`, or `engagement` . 
So if user entered {filter: {type: "internship"}}, then it is getting only interhshipengagement. 
And when {filter: {type: "language"}}, only languageengagement. When {filter: {type: "engagement or something"}}, both internship and language. 

2. Permission means user permission?

Reply:
1. No it's needs to be a list of products for the specific engagement given.
2. `canRead` and `canCreate` should not be hardcoded here https://github.com/SeedCompany/cord-api-v3/pull/296/files#diff-80606fe792a29f08bb910c50d206873cR500-R501

Reply:
I thought it is for list function. I got it.